File: src/test/java/shopping/persist/ProductRepositoryTest.java
package shopping.persist;

import static org.assertj.core.api.Assertions.assertThat;

import java.util.Optional;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.test.context.ContextConfiguration;
import shopping.domain.Product;

@ContextConfiguration(classes = {ProductRepository.class})
class ProductRepositoryTest extends JpaTest {

    @Autowired
    private ProductRepository productRepository;

    @Nested
    @DisplayName("findByProductName 메소드는")
    class findByProductName_method {

        private final Product product = new Product("제육덮밥", "/default-product.png", "10000");

        @Test
        @DisplayName("이름에 해당하는 Product를 반환한다.")
        void it_return_name_matched_product() {
            // given
            productRepository.saveProduct(product);

            // when
            Optional<Product> result = productRepository.findByProductName(product.getName());

            // then
            assertProduct(result);
        }

        @Test
        @DisplayName("이름에 해당하는 product가 없으면, Optional.empty를 반환한다.")
        void it_return_empty_optional_when_not_exists_product() {
            // when
            Optional<Product> result = productRepository.findByProductName(product.getName());

            // then
            assertThat(result).isNotPresent();
        }

        private void assertProduct(final Optional<Product> result) {
            assertThat(result).isPresent();
            assertThat(result.get().getId()).isNotNull();
            assertThat(result.get().getName()).isEqualTo(product.getName());
            assertThat(result.get().getImageUrl()).isEqualTo(product.getImageUrl());
            assertThat(result.get().getPrice()).isEqualTo(product.getPrice());
        }
    }

}

File: src/test/kotlin/shopping/ProductRepositoryTest.kt
package shopping

import io.kotest.matchers.shouldBe
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.jdbc.JdbcTest
import org.springframework.jdbc.core.JdbcTemplate
import shopping.domain.Product
import shopping.domain.ProductName

@JdbcTest
class ProductRepositoryTest {

    @Autowired
    lateinit var jdbcTemplate: JdbcTemplate

    private lateinit var productRepository: ProductRepository

    private val product = Product(
        id = 1,
        name = ProductName("아이스 카페 아메리카노 T"),
        price = 4500,
        imageUrl = "https://st.kakaocdn.net/product/gift/product/20231010111814_9a667f9eccc943648797925498bdd8a3.jpg",
    )

    @BeforeEach
    fun setUp() {
        productRepository = ProductRepository(jdbcTemplate)
    }

    @Test
    fun `상품을 추가하면 저장한다`() {
        // when
        val result = productRepository.saveProduct(product)

        // then
        result.name.name shouldBe "아이스 카페 아메리카노 T"
        result.price shouldBe 4500
    }

    @Test
    fun `상품을 저장하면 조회할 수 있다`() {
        // given
        productRepository.saveProduct(product)

        // when
        val result = productRepository.getById(product.id)

        // then
        result?.name?.name shouldBe "아이스 카페 아메리카노 T"
        result?.price shouldBe 4500
    }
}
